### PR TITLE
Debug draft for parsing multiple chracter command

### PR DESCRIPTION
**Pull request by gurgalex**
@NickHackman 

Draft of possible implementation for accepting `11c` to mean copy the current and further 10 images (11 total).

Unsure if this would be flexible for custom commands like:

Copy 11 images to specified destination path
`:11 copy dest`

### CODE REVIEW

**Comment by NickHackman** [comment]
We're on the same page, but in different places...

I'm on board with Modes being in `ui::State` that makes sense to me, along with the event matching 
loop in `ui.rs` as well

My flow of command is
run -> command_mode -> get input (Like the `run` event loop) -> process input and perform actions and switch back to normal mode.

I think having the program be able to switch modes in `run` (called from main) would make the most sense. I changed the name of the original `run` -> `run_normal_mode`. Without splitting these up we have a lot of code that's going to take place in `run`.

**Comment by gurgalex** [comment]
Good idea in splitting out the `run` fn into different types.

@NickHackman 
Seeing as how we need to settle on an implementation for the program as a whole. It doesn't seem time a good use of time for me to work on a competing implementation.

I'll work on making `delete` `copy` and `move` be able to take non-current indexes. The second part of #48

File: src/ui.rs [hunk 1]
```diff
@@ -36,16 +41,18 @@ pub enum Action {
     Move,
     /// Delete indicates the app should delete the image in response to this event
     Delete,
+    /// Put number on Screen for Numeric operations
+    NumericOp(Token),
+    /// Copy multiple images
+    CopyMultiple(&'a Vec<String>),
```
Comment (Davejkane) [inline src/ui.rs:47]: Not sure I like this design. I don't think `Copy` and `CopyMultiple` should be separate variants. I'd prefer copy took a number, and the default was 1. It shouldn't be its responsibility to do anything with a reference to a string vector.

I'm also not sure we need a string vector here. Logic could be: First time a number comes in, set Action to NumericOp with that number. Then there's two options, if a non number comes in, dispatch the action with the number we have, or if a number comes in, multiply the existing number by 10 and add the new one.

**Review by Davejkane — CHANGES_REQUESTED** [review]

**Comment by gurgalex** [comment]
Thanks for the early feedback.

That's much easier than I thought it would be to keep track of partial
numbers.

I think I'll rename `NumericOP` to `SwitchModeNumeric`. Which instructs the
program to switch to Numeric input entry mode for bulk operations.

I agree that I should modify the actions which make sense to perform in
bulk (copy, move, delete, next, prev, etc) to take a data in how many times
to perform.

Busy the next couple days, so probably around Sunday or Monday.

**Comment by Davejkane** [comment]
That's fine. It's a holiday weekend here in Denmark, so I probably wont have time to look at anything until Sunday or Monday as well. As always, no rush.

**Comment by NickHackman** [comment]
So I have an idea,

```rust
// src/ui,rs

pub enum Action<'a> {
    ...
    KeyboardInput<&'a str>, 
}

pub fn process_numeric_mode() -> Action<'a> {
// treat all numbers as KeyboardInput
// other actions just return
}

// src/program/mod.rs or src/program/numeric_mode.rs

// you could probably return a unsigned instead of String, if you want to parse in get_command
fn get_command(&self) -> (String, Action) {
    let numerical = String::new();
    let action: Action;
    // event loop
    // all KeyboardInput treat as valid append to numerical, render after every KeyboardInput on bottom right of infobar
    // break on all other Actions, set them to action
}

fn run_process_mode(&mut self) -> Result<(), String> {
    let (numerical_string, action) = self.get_command();
    // parse numerical string to number, it has to be valid otherwise we have a bug in get_command
    self.ui_state.mode = Mode::Normal; // return to Normal mode regardless of operation, for instance quit would change this to Exit mode
    match action {
         // perform options based on the amount to do and stuff
    }
}
```
This code is VERY close to what I've written, but we have different points to render at, different end conditions for our `get_command` etc...

Issues I'm facing, is when switching event loops it occasionally keeps the input in the event pump.
Furthermore rendering to the infobar is really hard, since the `render_screen()` function is the only one we can see, I think maybe decoupling them so we can update the info_bar then call a rerender all. Right now, my solution is sketchy just passing what to render to `render_screen()` and have it pass that down to infobar.

Also Rendering all Errors on the infobar would be a cool feature, in addition to terminal.

**Comment by gurgalex** [comment]
@NickHackman <https://github.com/NickHackman>
Could you put up your branch (on your GitHub account) so I can look at it.
We  could possibly find a solution together.

**Comment by Davejkane** [comment]
@NickHackman 👍 on rendering errors on the infobar. @gurgalex and I had discussed this previously. I'll write up an issue for it tomorrow though. No need to solve too many things on one branch. But always good to keep modularity in mind so that other problems will be easier to solve later on. Sorry if I didn't quite achieve that goal with the infobar implementation.

**Comment by gurgalex** [comment]
Closing in favor of #58

**Pull request closed without merging**